src/org/opensolaris/opengrok/web/api/v1/suggester/provider/service/impl/SuggesterServiceImpl.java

```diff
@@ -0,0 +1,382 @@
+import com.cronutils.model.definition.CronDefinitionBuilder;
+import com.cronutils.model.time.ExecutionTime;
+import com.cronutils.parser.CronParser;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.util.BytesRef;
+import org.opengrok.suggest.Suggester;
+import org.opengrok.suggest.Suggester.NamedIndexDir;
+import org.opengrok.suggest.Suggester.NamedIndexReader;
+import org.opengrok.suggest.Suggester.Suggestions;
+import org.opengrok.suggest.query.SuggesterQuery;
+import org.opensolaris.opengrok.configuration.Configuration;
+import org.opensolaris.opengrok.configuration.Project;
+import org.opensolaris.opengrok.configuration.RuntimeEnvironment;
+import org.opensolaris.opengrok.configuration.SuggesterConfig;
+import org.opensolaris.opengrok.configuration.SuperIndexSearcher;
+import org.opensolaris.opengrok.index.IndexDatabase;
+import org.opensolaris.opengrok.logger.LoggerFactory;
+import org.opensolaris.opengrok.web.api.v1.suggester.provider.service.SuggesterService;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Duration;
+import java.time.ZonedDateTime;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * Implementation of {@link SuggesterService}.
+ */
+public class SuggesterServiceImpl implements SuggesterService {
+
+    private static final Logger logger = LoggerFactory.getLogger(SuggesterServiceImpl.class);
+
+    private static SuggesterServiceImpl instance;
+
+    private Suggester suggester;
+
+    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+
+    private ScheduledFuture<?> future;
+
+    private final RuntimeEnvironment env = RuntimeEnvironment.getInstance();
+
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+
+    private SuggesterServiceImpl() {
+    }
+
+    public static SuggesterService getInstance() {
+        if (instance == null) {
+            instance = new SuggesterServiceImpl();
+            instance.initSuggester();
+        }
+        return instance;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Suggestions getSuggestions(
+            final Collection<String> projects,
+            final SuggesterQuery suggesterQuery,
+            final Query query
+    ) {
+        List<SuperIndexSearcher> superIndexSearchers = new LinkedList<>();
+        lock.readLock().lock();
```

Comment:
why it is necessary to lock this block ?

Reply:
Because if suggester is being enabled/disabled or totally refreshed then in `refresh()` method I close the old `suggester`, set it to `null` and then I reinitialize it from scratch. If that were to happen in the middle of the invocation of this method, NPE could be possibly thrown or closed `suggester` could be invoked.